File: prep/implementation/practice_iterations/merge_intervals_2/conflicting_appointments.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

// O(nlogn) / O(n)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    int s, e;
    for (int i = 0; i < n; i++) {
      cin >> s >> e;
      a.push_back(Interval(s, e));
    }

    sort(a.begin(), a.end(),
         [](auto &a, auto &b) { return a.start <= b.start; });

    bool can_attend = true;
    for (int i = 1; i < n; i++) {
      if (a[i].start < a[i - 1].end) {
        can_attend = false;
        break;
      }
    }

    cout << (can_attend ? "true" : "false") << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/merge_intervals_4/ConflictingAppointments.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(nlogn) / O(1)
public class ConflictingAppointments {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        a.sort((x, y) -> Integer.compare(x.start, y.start));

        boolean canAttend = true;
        for (int i = 1; i < n; i++) {
          if (hasConflict(a.get(i), a.get(i - 1))) {
            canAttend = false;
            break;
          }
        }

        System.out.println(canAttend);
      }
    }
  }

  private static boolean hasConflict(Interval a, Interval b) {
    return (a.start >= b.start && a.start < b.end) ||
        (b.start >= a.start && b.start < a.end);
  }
}